File: worker/src/lib/cleanup.ts
import type { Env } from "./types";

const RETENTION_DAYS = 35;

// wallet_details gets its own much shorter retention — it hit D1's 500MB
// size cap twice now (2026-07-26, then again 2026-08-08 with
// wallet_details at 2,073,708 rows / ~450MB of ~477MB total, still sitting
// right at the then-15-day boundary) because it stores every individual
// bet/bonus event, scaling with total bet COUNT rather than with
// (users x games). The 2026-08-08 fix moved Game Activity's "15days"
// period (Top Games/Highest Bet/Roller Active) onto wallet_daily_agg, the
// same rollup "month" already used — see the gameplaySource/gameplayBlock
// comments in index.ts. That leaves day/week as the only periods still on
// raw wallet_details, and those only ever need 7 days back
// (addDaysGA(anchorDate, -6), inclusive of anchorDate) — plus Search
// User's 2/7-day gameplay panels and Suspicious Withdrawals' 3-day games
// count, all comfortably under 7. Bonus Claims never had a 15days period
// (day/week/month only), so it was unaffected either way.
const WALLET_DETAILS_RETENTION_DAYS = 7;
const WALLET_DAILY_AGG_RETENTION_DAYS = 40;

// Every table in the Daily Records DB is a rolling window, not permanent
// storage — that's the whole point of splitting it from the Master DB.
// Each table archives to R2 (when available) before deleting, same
// optional-R2 pattern as upload.ts: if R2 isn't enabled yet, cleanup still
// runs, it just skips the archive.
const TABLES: { name: string; dateColumn: string; retentionDays: number }[] = [
  { name: "sync_runs", dateColumn: "started_at", retentionDays: RETENTION_DAYS },
  { name: "deposits", dateColumn: "synced_at", retentionDays: RETENTION_DAYS },
  { name: "withdrawals", dateColumn: "synced_at", retentionDays: RETENTION_DAYS },
  { name: "wallet_details", dateColumn: "synced_at", retentionDays: WALLET_DETAILS_RETENTION_DAYS },
  // dateColumn is "d" (the calendar day the row summarizes), not
  // synced_at — wallet_daily_agg rows are recomputed in place each run,
  // not append-only, so there's no separate sync timestamp to purge by.
  { name: "wallet_daily_agg", dateColumn: "d", retentionDays: WALLET_DAILY_AGG_RETENTION_DAYS },
];

export async function cleanupOldSyncRuns(env: Env): Promise<{ archived: number; deleted: number; perTable: Record<string, number> }> {
  const perTable: Record<string, number> = {};
  let archived = 0;
  let deleted = 0;

  for (const { name, dateColumn, retentionDays } of TABLES) {
    const cutoff = new Date(Date.now() - retentionDays * 24 * 60 * 60 * 1000).toISOString();
    const old = await env.daily_records_db
      .prepare(`SELECT * FROM ${name} WHERE ${dateColumn} < ?`)
      .bind(cutoff)
      .all();

    perTable[name] = old.results.length;
    if (old.results.length === 0) continue;

    if (env.UPLOADS) {
      const key = `archived-logs/${name}-${new Date().toISOString().slice(0, 10)}.json`;
      try {
        await env.UPLOADS.put(key, JSON.stringify(old.results));
        archived += old.results.length;
      } catch (err) {
        console.error(`Archive to R2 failed for ${name} (non-fatal, proceeding with delete):`, err);
      }
    }

    await env.daily_records_db.prepare(`DELETE FROM ${name} WHERE ${dateColumn} < ?`).bind(cutoff).run();
    deleted += old.results.length;
  }

  return { archived, deleted, perTable };
}

File: etl/cleanup.py
"""Entry point: purges Daily Records DB rows past their retention window.
Mirrors worker/src/lib/cleanup.ts's tiered retention — this file used to
apply a flat 35 days to every table including wallet_details, which was
stale/wrong: wallet_details actually needs a much shorter window (see
cleanup.ts's comment) since it scales with total bet COUNT, not (users x
games), and hit D1's 500MB size cap twice (2026-07-26, 2026-08-08) sitting
at the old, too-long retention figure. In practice worker-upload's own
daily cron already runs cleanupOldSyncRuns() with the correct tiered
retention (see worker-upload/wrangler.jsonc), so this script is currently
redundant belt-and-suspenders rather than the primary enforcement — but it
needs to match, not silently apply a wrong/longer retention, in case that
cron is ever the one that's down. R2 archiving is included since Python has
no issue talking to R2's S3-compatible API directly; skipped automatically
if R2_BUCKET isn't set."""
import os
from datetime import datetime, timedelta, timezone

import cf_client

DAILY_DB_ID = os.environ["DAILY_DB_ID"]
RETENTION_DAYS = 35
WALLET_DETAILS_RETENTION_DAYS = 7
WALLET_DAILY_AGG_RETENTION_DAYS = 40

TABLES = [
    ("sync_runs", "started_at", RETENTION_DAYS),
    ("deposits", "synced_at", RETENTION_DAYS),
    ("withdrawals", "synced_at", RETENTION_DAYS),
    ("wallet_details", "synced_at", WALLET_DETAILS_RETENTION_DAYS),
    # dateColumn is "d" (the calendar day the row summarizes), not
    # synced_at — wallet_daily_agg rows are recomputed in place each run,
    # not append-only, so there's no separate sync timestamp to purge by.
    ("wallet_daily_agg", "d", WALLET_DAILY_AGG_RETENTION_DAYS),
]


def main():
    now = datetime.now(timezone.utc)
    total_deleted = 0

    for table, date_column, retention_days in TABLES:
        cutoff = (now - timedelta(days=retention_days)).isoformat()
        old_rows = cf_client.d1_query(DAILY_DB_ID, f"SELECT COUNT(*) as c FROM {table} WHERE {date_column} < ?", [cutoff])
        count = old_rows[0]["c"] if old_rows else 0
        if count == 0:
            print(f"[{table}] nothing older than {retention_days} days")
            continue

        # R2 archiving intentionally omitted here until R2 is enabled in the
        # dashboard (see UPLOADS binding notes in the Workers version) —
        # add a boto3-based archive step here once a bucket exists.

        cf_client.d1_query(DAILY_DB_ID, f"DELETE FROM {table} WHERE {date_column} < ?", [cutoff])
        total_deleted += count
        print(f"[{table}] deleted {count} rows older than {retention_days} days")

    print(f"cleanup complete — {total_deleted} total rows deleted")


if __name__ == "__main__":
    main()
